sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InferWindowGroupLimit.scala

```diff
@@ -84,11 +84,11 @@ object InferWindowGroupLimit extends Rule[LogicalPlan] with PredicateHelper {
           }
           // Pick a rank-like function with the smallest limit
           selectedLimits.minBy(_._1) match {
-            case (limit, rankLikeFunction) if limit <= conf.windowGroupLimitThreshold =>
+            case (limit, rankLikeFunction) if limit <= conf.windowGroupLimitThreshold &&
+              child.maxRows.forall(_ > limit) =>
```

Comment:
Is this inevitable when we move the rule?

Reply:
This is to ensure the rule's idempotence.

Reply:
So, before this reverting, the idempotence was ensured?

Reply:
In previous, the idempotence not be exposed.

Reply:
I'm also curious about why we suddenly detect idempotence issue. The batch is still Once.

Reply:
In previous, the batch only contains `InferWindowGroupLimit`. After extracts one Limit between Filter and Window, so the plan not meet `InferWindowGroupLimit` again.

This PR added other rules into the same batch, so the Limit extracted by `InferWindowGroupLimit`, will be pushed down to Window. So the plan still meet `InferWindowGroupLimit`.